src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -0,0 +1,363 @@
+
+        # All pairs (i,j) where i<j
+        indices = np.arange(n_ims)
+        pairs = [(i, j) for idx, i in enumerate(indices) for j in indices[idx + 1 :]]
+
+        # Populate upper triangle of V with vijs
+        for idx, (i, j) in enumerate(pairs):
+            V[3 * i : 3 * (i + 1), 3 * j : 3 * (j + 1)] = vijs[idx]
+
+        # Populate lower triangle of V with vjis, where vji = vij^T
+        V = V + V.T
+
+        # Populate diagonal of V with viis
+        for i in range(n_ims):
+            V[3 * i : 3 * (i + 1), 3 * i : 3 * (i + 1)] = viis[i]
+
+        # In a clean setting V is of rank 1 and its eigenvector is the concatenation
+        # of the third rows of all rotation matrices.
+        # In the noisy setting we use the eigenvector corresponding to the leading eigenvalue
+        val, vec = eig(V)
+        lead_idx = np.argsort(val)[-1]
+        lead_vec = vec[:, lead_idx]
+
+        vis = lead_vec.reshape((n_ims, 3))
+        for i in range(n_ims):
+            vis[i] = vis[i] / norm(vis[i])
+
+        return vis
+
+    def _estimate_inplane_rotations(
+        self, pf, vis, inplane_rot_res, max_shift, shift_step
+    ):
+        # return rots
+        pass
+
+    #################################################
+    # Secondary Methods for computing outer product #
+    #################################################
+
+    def _self_clmatrix_c3_c4(
+        self, pf, n_symm, max_shift, shift_step, is_handle_equator_ims
+    ):
+        # return sclmatrix
+        pass
+
+    def _estimate_all_Riis_c3_c4(n_symm, sclmatrix, n_theta):
+        # return Riis
+        pass
+
+    def _estimate_all_Rijs_c3_c4(n_symm, clmatrix, n_theta):
+        # return Rijs
+        pass
+
+    def local_sync_J_c3_c4(n_symm, Rijs, Riis):
+        # return vijs, viis
+        pass
+
+    #######################################
+    # Secondary Methods for Global J Sync #
+    #######################################
+
+    def _J_sync_power_method(self, vijs):
+        """
+        Calculate the leading eigenvector of the J-synchronization matrix
+        using the power method.
+
+        As the J-synchronization matrix is of size (N choose 2)x(N choose 2), we
+        use the power method to the compute the eigenvalues and eigenvectors,
+        while constructing the matrix on-the-fly.
+
+        :param vijs: nchoose2x3x3 array of estimates of relative orientation matrices.
+
+        :return: Array of length N-choose-2 where the i-th entry indicates if vijs[i]
+        should be J-conjugated or not to achieve global handedness consistency. This array
+        consists only of +1 and -1.
+        """
+
+        n_vijs = vijs.shape[0]
+        nchoose2 = (1 + np.sqrt(1 + 8 * n_vijs)) / 2
+        assert nchoose2 == int(nchoose2), "There must be n_ims-choose-2 vijs."
+        # assert n_eigs > 0, "n_eigs must be a positive integer."
+
+        epsilon = 1e-2
+        max_iters = 100
```

Comment:
I think this should probably be a class or method option. Also is there any link between epsilon and required iterations, like any accuracy guarantees for the power method?